plugins/kubernetes/app/models/watchers/deploy_watcher.rb

```diff
@@ -1,119 +1,265 @@
     end
 
-    def update_timeout(pod_event)
-      unless pod_event.deleted?
-        @pod_timer.reset if pod_event.pod.ready?
+    def update_release_doc(release_doc, live_pods)
+      reset_timeout
+
+      if release_doc.live_replicas_changed?(live_pods)
+        release_doc.update_status(live_pods)
+        release_doc.update_replica_count(live_pods)
+        release_doc.update_release
+        yield
       end
     end
 
-    def end_deploy
-      @release.release_is_live!
-      info 'Deploy is live!'
-      terminate
+    def update_failed_release_doc(release_doc, live_pods)
+      if release_doc.live_replicas_changed?(live_pods)
+        release_doc.update_replica_count(live_pods)
+      end
+    end
+
+    # Involved object identifies the Kubernetes resource that triggered the event
+    def involved_object_for(notice)
+      notice.object.involvedObject
+    end
+
+    def cluster_for(cluster_id)
+      Kubernetes::Cluster.find(cluster_id)
+    end
+
+    def rcs
+      @rcs ||= {}
+    end
+
+    def rc_pods(rc_unique_id)
+      rcs[rc_unique_id] ||= {}
+      rcs[rc_unique_id]
+    end
+
+    def update_rc_pods(pod)
+      rc = rc_pods(pod.rc_unique_identifier)
+      rc[pod.name] = pod
+    end
+
+    def rc_pod(pod)
+      rc = rc_pods(pod.rc_unique_identifier)
+      rc[pod.name]
+    end
+
+    def release_ids_from_cluster
+      rcs.map { |_, pods| pods.map { |_, pod| pod.release_id } }.flatten.uniq
+    end
+
+    def last_release(project)
+      project.kubernetes_releases.last
+    end
+
+    def count_live_pods(rc_unique_identifier)
+      rc = rc_pods(rc_unique_identifier)
+      rc.reduce(0) { |count, (_, pod)| count += 1 if pod.live?; count }
+    end
+
+    def count_failed_pods(rc_unique_identifier)
+      rc = rc_pods(rc_unique_identifier)
+      rc.reduce(0) { |count, (_, pod)| count += 1 if failed?(pod); count }
+    end
+
+    def failed?(pod)
+      pod.respond_to?(:failed?) && pod.failed?
+    end
+
+    def reset_timeout
+      @pod_timer.reset
     end
 
     def pod_timeout
-      @release.release_failed!
+      last_release(project).fail!
       warn 'Deploy failed!'
       terminate
     end
 
-    def rc_pods(name)
-      @current_rcs[name] ||= {}
-      @current_rcs[name]
+    def deploy_finished?(project)
+      last_release(project).live? && project.kubernetes_releases.excluding(last_release(project).id).all?(&:dead?)
     end
 
-    def send_event(options)
-      base = {
-          project: @release.project.id,
-          build: @release.build.label,
-          release: @release.id
-      }
+    def on_termination
+      @pod_timer.cancel
+      info('Finished Watching Deployments!')
+    end
+
+    def terminate_watcher
+      info('Deploy finished!')
+      terminate
+    end
 
-      Rails.logger.info("[SSE] Sending: #{base.merge(options)}")
+    def send_event(release_doc, failed_pods)
+      info("[SSE] Sending: #{sse_event_data(release_doc, failed_pods)}")
```

Comment:
Added it back here because it might be useful to track deployments. It does not cause too much noise because the events are filtered.